core/common/lib/policy-engine-lib/src/test/java/org/eclipse/edc/policy/engine/PolicyEngineImplValidationTest.java

```diff
@@ -0,0 +1,295 @@
+import static org.eclipse.edc.policy.engine.spi.PolicyEngine.ALL_SCOPES;
+import static org.eclipse.edc.policy.model.Operator.EQ;
+import static org.junit.jupiter.params.provider.Arguments.of;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class PolicyEngineImplValidationTest {
+
+    private final RuleBindingRegistry bindingRegistry = new RuleBindingRegistryImpl();
+    private PolicyEngine policyEngine;
+
+
+    @BeforeEach
+    void setUp() {
+        policyEngine = new PolicyEngineImpl(new ScopeFilter(bindingRegistry), new RuleValidator(bindingRegistry));
+    }
+
+    @Test
+    void validateEmptyPolicy() {
+        var emptyPolicy = Policy.Builder.newInstance().build();
+
+        var result = policyEngine.validate(emptyPolicy);
+
+        assertThat(result).isSucceeded();
+    }
+
+    @Test
+    void validate_whenKeyNotBoundInTheRegistryAndToFunctions() {
+
+        var left = new LiteralExpression("foo");
+        var right = new LiteralExpression("bar");
+        var constraint = AtomicConstraint.Builder.newInstance().leftExpression(left).operator(EQ).rightExpression(right).build();
+        var permission = Permission.Builder.newInstance().constraint(constraint).build();
+        var policy = Policy.Builder.newInstance().permission(permission).build();
+        policyEngine.registerFunction(ALL_SCOPES, Duty.class, "foo", (op, rv, r, ctx) -> true);
+        policyEngine.registerFunction(ALL_SCOPES, Prohibition.class, "foo", (op, rv, r, ctx) -> true);
+
+        var result = policyEngine.validate(policy);
+
+        // The foo key is not bound nor to function nor in the RuleBindingRegistry
+        assertThat(result).isFailed().messages().hasSize(2)
+                .anyMatch(s -> s.startsWith("leftOperand 'foo' is not bound to any scopes"))
+                .anyMatch(s -> s.startsWith("left operand 'foo' is not bound to any functions"));
+
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(PolicyProvider.class)
+    void validate_whenKeyIsNotBoundInTheRegistry(Policy policy, Class<Rule> ruleClass, String key) {
+
+        policyEngine.registerFunction(ALL_SCOPES, ruleClass, key, (op, rv, duty, ctx) -> true);
+
+        var result = policyEngine.validate(policy);
+
+        // The input key is not bound in the RuleBindingRegistry
+        assertThat(result).isFailed().messages().hasSize(1)
+                .anyMatch(s -> s.startsWith("leftOperand '%s' is not bound to any scopes".formatted(key)));
+
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(PolicyProvider.class)
+    void validate(Policy policy, Class<Rule> ruleClass, String key) {
+
+
+        bindingRegistry.bind(key, ALL_SCOPES);
+        policyEngine.registerFunction(ALL_SCOPES, ruleClass, key, (op, rv, duty, ctx) -> true);
+
+        var result = policyEngine.validate(policy);
+
+        // The input key is not bound in the RuleBindingRegistry
```

Comment:
this comment doesn't look correct, am I right?

Reply:
yeah copy and paste :)